stan/math/prim/fun/vec_concat.hpp

```diff
@@ -13,24 +13,25 @@ namespace math {
  * @param v1 Events on the OpenCL event stack.
  * @return returns input vector.
  */
-template <typename T>
-inline const std::vector<T>& vec_concat(const std::vector<T>& v1) {
-  return v1;
+template <typename Vec>
+inline decltype(auto) vec_concat(Vec&& v1) {
+  return std::forward<Vec>(v1);
 }
 
 /**
  * Get the event stack from a vector of events and other arguments.
  *
- * @tparam T type of elements in the array
- * @tparam Args types for variadic arguments
+ * @tparam Vec type of first input standard vector
+ * @tparam VecArgs types for standard vector variadic arguments
  * @param v1 event stack to roll up
  * @param args variadic arguments passed down to the next recursion
  * @return Vector of OpenCL events
  */
-template <typename T, typename... Args>
-inline const std::vector<T> vec_concat(const std::vector<T>& v1,
-                                       const Args... args) {
-  std::vector<T> vec = vec_concat(args...);
+template <typename Vec, typename... VecArgs,
+          require_all_std_vector_t<Vec, VecArgs...>* = nullptr>
+inline auto vec_concat(Vec&& v1, VecArgs&&... args) {
+  std::vector<value_type_t<Vec>> vec
+      = vec_concat(std::forward<VecArgs>(args)...);
```

Comment:
[optional] This is a bit inefficient. The result vector `vec` should reserve its final size in advance.